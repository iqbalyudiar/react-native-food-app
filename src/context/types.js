export const GET_RESTAURANTS = "get_restaurants";
export const ADD_TO_CART = "add_to_cart";
export const REMOVE_FROM_CART = "remove_from_cart";
