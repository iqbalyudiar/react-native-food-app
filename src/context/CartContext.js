import React, { useReducer } from "react";
import { ADD_TO_CART, REMOVE_FROM_CART } from "./types";

const initialState = {
  carts: [],
};

const cartReducer = (state, action) => {
  const found = state.carts.some((cart) => cart.id === action.payload.id);
  switch (action.type) {
    case ADD_TO_CART:
      if (found) {
        const data = state.carts.map((cart) =>
          cart.id === action.payload.id
            ? { ...cart, quantity: cart.quantity + 1 }
            : cart
        );
        return { ...state, carts: data };
      }

      return {
        ...state,
        carts: [...state.carts, ...[{ ...action.payload, quantity: 1 }]],
      };
    case REMOVE_FROM_CART:
      if (found) {
        const newCarts = state.carts
          .map((cart) =>
            cart.id === action.payload.id && cart.quantity > 0
              ? { ...cart, quantity: cart.quantity - 1 }
              : cart
          )
          .filter((item) => item.quantity > 0);

        return { ...state, carts: newCarts };
      }

      return { ...state };

    default:
      return state;
  }
};

export const Context = React.createContext();

export const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);

  const addToCart = (food) => {
    dispatch({ type: ADD_TO_CART, payload: food });
  };

  const removeFromCart = (id) => {
    dispatch({ type: REMOVE_FROM_CART, payload: { id } });
  };

  const actions = { addToCart, removeFromCart };

  const value = { state, ...actions };

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
