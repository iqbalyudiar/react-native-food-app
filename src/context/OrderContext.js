import React, { useReducer } from "react";
import foodApi from "../api/food";

const initialState = {
  orders: [],
};

const orderReducer = (state, action) => {
  switch (action.type) {
    default:
      return state;
  }
};

export const Context = React.createContext();

export const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(orderReducer, initialState);

  const addOrder = async ({ carts, restaurant, totalPrice }, callback) => {
    try {
      const params = {
        id: Math.floor(Math.random() * 99999),
        created_at: Date.now(),
        items: carts,
        restaurant,
        totalPrice,
      };

      await foodApi.post("/orders", params);
      if (callback) {
        callback();
      }
    } catch (error) {
      console.error(error.message);
    }
  };
  const getOrders = async () => {};

  const actions = { addOrder, getOrders };

  const value = { state, ...actions };

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
