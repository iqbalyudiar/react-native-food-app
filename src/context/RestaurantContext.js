import React, { useReducer } from "react";

const initialState = {
  restaurant: [],
};

const restaurantReducer = (state, action) => {
  switch (action.type) {
    default:
      return state;
  }
};

export const Context = React.createContext();

export const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(restaurantReducer, initialState);

  const getRestaurants = () => {};

  const value = { state, getRestaurants };

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
