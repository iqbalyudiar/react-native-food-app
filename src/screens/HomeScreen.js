import { StyleSheet, Text, View } from "react-native";
import { Button } from "@rneui/base";
import React from "react";

const HomeScreen = ({ navigation }) => {
  return (
    <View>
      <Text>HomeScreen</Text>
      <Button title="Go to cart" onPress={() => navigation.navigate("Cart")} />
    </View>
  );
};

export default HomeScreen;

const styles = StyleSheet.create({});
