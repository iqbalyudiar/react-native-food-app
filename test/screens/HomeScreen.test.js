import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react-native";
import HomeScreen from "~/screens/HomeScreen";
import { Provider as RestaurantProvider } from "~/context/RestaurantContext";

import { NavigationContext } from "@react-navigation/native";

afterAll(() => {
  jest.restoreAllMocks();
});

it("render default", async () => {
  const navContextValue = {
    isFocused: () => true,
    addListener: jest.fn(() => jest.fn()),
  };
  const navigation = {
    navigate: jest.fn(),
  };

  const mockData = {
    id: 1,
    name: "KFC",
  };
  const wrapper = render(
    <NavigationContext.Provider value={navContextValue}>
      <RestaurantProvider>
        <HomeScreen navigation={navigation} />
      </RestaurantProvider>
    </NavigationContext.Provider>
  );

  await waitFor(() => {
    expect(wrapper.getByTestId("home-screen-navigate-1")).toBeTruthy();
  });
  fireEvent.press(wrapper.getByTestId("home-screen-navigate-1"), mockData);

  expect(wrapper).toMatchSnapshot();
});
