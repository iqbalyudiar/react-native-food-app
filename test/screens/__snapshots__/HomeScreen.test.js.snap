// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<View>
  <RCTScrollView
    data={
      Array [
        Object {
          "foods": Array [
            Object {
              "id": 1,
              "name": "Chicken",
              "price": 20000,
            },
            Object {
              "id": 2,
              "name": "Burger",
              "price": 30000,
            },
          ],
          "id": 1,
          "logo": "https://seeklogo.com/images/M/mc-donald-s-logo-8568BCE03C-seeklogo.com.png",
          "name": "Mc Donalds",
          "rating": "4.8",
        },
        Object {
          "foods": Array [
            Object {
              "id": 1,
              "name": "Chicken",
              "price": 20000,
            },
            Object {
              "id": 2,
              "name": "Burger",
              "price": 30000,
            },
          ],
          "id": 2,
          "logo": "https://seeklogo.com/images/K/kfc-new-logo-72E6348046-seeklogo.com.png",
          "name": "KFC",
          "rating": "4.8",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="home-screen-navigate-1"
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#ffffff",
                  "borderBottomWidth": 0.5,
                  "borderColor": "#bcbbc1",
                  "flexDirection": "row",
                  "padding": 14,
                }
              }
              testID="RNE__LISTITEM__padView"
            >
              <View
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "height": 34,
                    "width": 34,
                  }
                }
              >
                <View
                  accessibilityIgnoresInvertColors={true}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "flex": 1,
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }
                >
                  <Image
                    onLoad={[Function]}
                    source={
                      Object {
                        "uri": "https://seeklogo.com/images/M/mc-donald-s-logo-8568BCE03C-seeklogo.com.png",
                      }
                    }
                    style={
                      Object {
                        "bottom": 0,
                        "flex": 1,
                        "height": undefined,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "width": undefined,
                      }
                    }
                    testID="RNE__Avatar__Image"
                    transitionDuration={360}
                  />
                  <View
                    accessibilityElementsHidden={true}
                    collapsable={false}
                    importantForAccessibility="no-hide-descendants"
                    pointerEvents="none"
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "opacity": 1,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#bdbdbd",
                          "flex": 1,
                          "height": undefined,
                          "justifyContent": "center",
                          "width": undefined,
                        }
                      }
                      testID="RNE__Image__placeholder"
                    >
                      <Text
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#ffffff",
                            "fontSize": 17,
                            "textAlign": "center",
                          }
                        }
                      >
                        M
                      </Text>
                    </View>
                  </View>
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "height": undefined,
                        "width": undefined,
                      }
                    }
                    testID="RNE__Image__children__container"
                  />
                </View>
              </View>
              <View
                style={
                  Object {
                    "paddingLeft": 16,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "flex-start",
                      "flex": 1,
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  accessibilityRole="text"
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "color": "#242424",
                      "fontSize": 17,
                    }
                  }
                  testID="listItemTitle"
                >
                  Mc Donalds
                </Text>
              </View>
              <View
                style={
                  Object {
                    "paddingLeft": 16,
                  }
                }
              />
              <View
                style={
                  Object {
                    "alignSelf": "center",
                    "overflow": "hidden",
                  }
                }
                testID="RNE__ICON__CONTAINER"
              >
                <View
                  accessibilityRole="button"
                  android_ripple={
                    Object {
                      "borderless": false,
                      "color": "rgba(0, 0, 0, 0.3)",
                      "radius": -5,
                    }
                  }
                  disabled={false}
                  testID="RNE__ICON__CONTAINER_ACTION"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    testID="RNE__ICON"
                  >
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "#D1D1D6",
                            "fontSize": 16,
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                          Object {
                            "fontFamily": "ionicons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                      testID="RNE__ICON__Component"
                    >
                      ïˆ˜
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="home-screen-navigate-2"
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#ffffff",
                  "borderBottomWidth": 0.5,
                  "borderColor": "#bcbbc1",
                  "flexDirection": "row",
                  "padding": 14,
                }
              }
              testID="RNE__LISTITEM__padView"
            >
              <View
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "height": 34,
                    "width": 34,
                  }
                }
              >
                <View
                  accessibilityIgnoresInvertColors={true}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "flex": 1,
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }
                >
                  <Image
                    onLoad={[Function]}
                    source={
                      Object {
                        "uri": "https://seeklogo.com/images/K/kfc-new-logo-72E6348046-seeklogo.com.png",
                      }
                    }
                    style={
                      Object {
                        "bottom": 0,
                        "flex": 1,
                        "height": undefined,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "width": undefined,
                      }
                    }
                    testID="RNE__Avatar__Image"
                    transitionDuration={360}
                  />
                  <View
                    accessibilityElementsHidden={true}
                    collapsable={false}
                    importantForAccessibility="no-hide-descendants"
                    pointerEvents="none"
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "opacity": 1,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#bdbdbd",
                          "flex": 1,
                          "height": undefined,
                          "justifyContent": "center",
                          "width": undefined,
                        }
                      }
                      testID="RNE__Image__placeholder"
                    >
                      <Text
                        style={
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#ffffff",
                            "fontSize": 17,
                            "textAlign": "center",
                          }
                        }
                      >
                        K
                      </Text>
                    </View>
                  </View>
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "height": undefined,
                        "width": undefined,
                      }
                    }
                    testID="RNE__Image__children__container"
                  />
                </View>
              </View>
              <View
                style={
                  Object {
                    "paddingLeft": 16,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "flex-start",
                      "flex": 1,
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  accessibilityRole="text"
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "color": "#242424",
                      "fontSize": 17,
                    }
                  }
                  testID="listItemTitle"
                >
                  KFC
                </Text>
              </View>
              <View
                style={
                  Object {
                    "paddingLeft": 16,
                  }
                }
              />
              <View
                style={
                  Object {
                    "alignSelf": "center",
                    "overflow": "hidden",
                  }
                }
                testID="RNE__ICON__CONTAINER"
              >
                <View
                  accessibilityRole="button"
                  android_ripple={
                    Object {
                      "borderless": false,
                      "color": "rgba(0, 0, 0, 0.3)",
                      "radius": -5,
                    }
                  }
                  disabled={false}
                  testID="RNE__ICON__CONTAINER_ACTION"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    testID="RNE__ICON"
                  >
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "#D1D1D6",
                            "fontSize": 16,
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                          Object {
                            "fontFamily": "ionicons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                      testID="RNE__ICON__Component"
                    >
                      ïˆ˜
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
